### ISSUE TITLE
kube-proxy is evicted even with ExperimentalCriticalPodAnnotation enabled

### ISSUE THREAD

**Issue by vainu-arto** [kubernetes/kubernetes#58379]
/kind bug

On my kops deployed cluster kube-proxy is deployed as a static pod and has scheduler.alpha.kubernetes.io/critical-pod= in annotations. The kubelet has these feature gates enabled:
--feature-gates=ExperimentalCriticalPodAnnotation=true,PodPriority=true

**What happened**:

kube-proxy is constantly evicted due to pressure from user pods that have podpriority set to above zero.

**What you expected to happen**:

Not have kube-proxy evicted.

**Environment**:

This is on Kubernetes v1.8.6 deployed by kops 1.8.0, with podpriority enabled and in use. The cluster is running on AWS and using Debian Stretch as the base os.

**Comment by MrHohn** [comment]
@kubernetes/sig-scheduling-bugs

**Comment by vainu-arto** [comment]
I'm not currently set up to have easy testability on different versions, I'll see what I can do about that..

**Comment by k82cn** [comment]
@vainu-arto , np, I can test it with latest version if you did not have ready env :).

**Comment by bsalamat** [comment]
This is WAI. If you enable Pod priority and preemption and some of your Pods have larger than zero priority, they could cause lower priority pods to get evicted. In order to solve this issue, you need to set priority class name of kube-proxy to `system-node-critical`. `system-node-critical` is a predefined priority class and is the highest priority level.

**Comment by vainu-arto** [comment]
What is the relationship between system-node-critical (which seems to be undocumented) and SYSTEM (which is documented at https://kubernetes.io/docs/api-reference/v1.9/#podspec-v1-core )?

In either case, kops doesn't appear to have a way to change this (the kube-proxy static pod manifest seems to be hard-coded in nodeup).

Is setting PriorityClassName an error if PodPriority isn't enabled (I assume so)?

**Comment by vainu-arto** [comment]
Manually adding priorityClassName set to either system-node-critical or SYSTEM to the static pod manifest stops the pod from starting.

With SYSTEM an error message is returned:
" Can't process manifest file "/etc/kubernetes/manifests/kube-proxy.manifest": invalid pod: [spec.priorityClassName: Invalid value: "SYSTEM": a DNS-1123 subdomain must consist of lower case alphanumeric characters, '-' or '.', and must start and end with an alphanumeric character (e.g. 'example.com', regex used for validation is '[a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*')]"

With system-node-critical there is no discernible error message, but the kube-proxy pod fails to start.

**Comment by bsalamat** [comment]
> What is the relationship between system-node-critical (which seems to be undocumented) and SYSTEM (which is documented at https://kubernetes.io/docs/api-reference/v1.9/#podspec-v1-core )?

The SYSTEM in that document is comprised of two priority classes: `system-node-critical` and `system-cluster-critical`. `system-node-critical` is the highest priority. (I should update that document.)

> In either case, kops doesn't appear to have a way to change this (the kube-proxy static pod manifest seems to be hard-coded in nodeup).

That's possible. I am not familiar enough with kops to provide guidance, but I know that you can run kube-proxy as a DaemonSet as well. I am not sure if this is possible with kops though.

> Is setting PriorityClassName an error if PodPriority isn't enabled (I assume so)?

Yes, it is.

**Comment by bsalamat** [comment]
> With system-node-critical there is no discernible error message, but the kube-proxy pod fails to start.

Can you get "kubectl describe pod" for the kube-proxy and send the information?

**Comment by bsalamat** [comment]
> So ok, setting priorityClassName: system-node-critical seems like the correct fix but implementing it seems somewhat difficult with kops.

Thanks for verifying that it works and also thanks for trying priority and bringing it to our attention that kops may has an issue with it.

**Comment by vainu-arto** [comment]
So setting podpriorityclass does actually work, but it doesn't fix the issue I'm seeing. Even with priorityClassName: system-node-critical the kube-proxy pods are constantly evicted from their nodes.

**Comment by vainu-arto** [comment]
And here is the smoking gun. For a kube-dns pod created out of a deployment with priorityclass set:
  priority: 2000000000
  priorityClassName: system-cluster-critical

For the static kube-proxy pod:
  priorityClassName: system-node-critical

So apparently kubelet doesn't set priority when it creates the static pod, and it needs to be set manually. In that case there is no reason to set priorityClassName, is there?

**Comment by k82cn** [comment]
> it needs to be set manually

nop, it should be handled by admission controller; Bobby and me just talked it this afternoon :). It's interesting that kube-dns was updated but kube-proxy not, let me check it.

**Comment by vainu-arto** [comment]
Does a static pod pass through the mentioned admission controller?

**Comment by k82cn** [comment]
@vainu-arto , should be; let me double check the logic when kubelet creating it.

**Comment by k82cn** [comment]
wow, we skip mirror pod in admission controller [here](https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/priority/admission.go#L159).

**Comment by ravisantoshgudimetla** [comment]
@k82cn @vainu-arto I have just started playing around with priorities and it seems the problem is indeed with this line in admission controller(https://github.com/kubernetes/kubernetes/blob/master/plugin/pkg/admission/priority/admission.go#L159) because of which we are not setting priority for it

**Comment by ravisantoshgudimetla** [comment]
Oops @k82cn you bet me it..

**Comment by bsalamat** [comment]
cc/ @dashpole for kubelet evictions

**Comment by dashpole** [comment]
@vainu-arto what does the event look like when it is evicted?  Is the reason `Evicted`, `Preempting`, or something else?

**Comment by vainu-arto** [comment]
I did not make notes of these, and am not right now in a position to retest. This should be relatively simple to reproduce in any test clusters you have, though.

**Comment by dashpole** [comment]
@vainu-arto do you have logs from the node where the evictions took place?  I would like to verify that the node started up with the feature gate set, and figure out which resource caused the eviction.

**Comment by vainu-arto** [comment]
I did not store the logs, sadly. On my cluster kubelet is started like this:
/usr/local/bin/kubelet --allow-privileged=true --cgroup-root=/ --cloud-provider=aws --cluster-dns=<redacted> --cluster-domain=cluster.local --enable-custom-metrics=true --enable-debugging-handlers=true --eviction-hard=memory.available<100Mi,nodefs.available<10%,nodefs.inodesFree<5%,imagefs.available<10%,imagefs.inodesFree<5% --feature-gates=ExperimentalCriticalPodAnnotation=true,PodPriority=true --hostname-override=<redacted> --kubeconfig=/var/lib/kubelet/kubeconfig --network-plugin=cni --node-labels=awstype=ondemand,kops.k8s.io/instancegroup=nodes,kubernetes.io/role=node,node-role.kubernetes.io/node= --non-masquerade-cidr=<redacted> --pod-infra-container-image=gcr.io/google_containers/pause-amd64:3.0 --pod-manifest-path=/etc/kubernetes/manifests --register-schedulable=true --require-kubeconfig=true --v=2 --cni-bin-dir=/opt/cni/bin/ --cni-conf-dir=/etc/cni/net.d/

IIRC the changes from kops defaults are enabling podpriority and custom metrics (kops enables criticalannotation by default since a lot of things rely on it, sadly it doesn't really work if priorities are enabled).

I assume the evictions happened simply because there were hundreds of tasks with a priority set pending (this was running a workload that generates several hundred jobs, and we are trying to use priorities to (eh) prioritise them). Since kube-proxy doesn't have a priority set (even though priorityclassname was specified) due to this bug anything with a priority causes it to be evicted.

**Comment by ravisantoshgudimetla** [comment]
> I am not convinced this is a kubelet eviction. It seems more likely that the pod was rescheduled. The logic for enforcing that critical pods are not evicted is fairly simple, and it doesn't depend on pod pri

@dashpole I agree with you. I will set up a cluster to test this out sometime this week.

The other question that I have is will static pods be able evict other running pods on node similar to critical pods. I believe it is not the case currently. Shouldn't they evict?

**Comment by dashpole** [comment]
>  will static pods be able evict other running pods on node similar to critical pods?

I am not 100% sure what you mean by static pods evicting other pods, but ill take a stab.  If you create a static pod with the critical pod annotation in the kube-system namespace, and there are no free resources on the node to schedule the pod, then the kubelet will preempt a pod on the node to make room for it.  This is only for critical pods.

**Comment by ravisantoshgudimetla** [comment]
@dashpole Thanks for the explanation. My question is what if static pod is created without critical pod annotation? My understanding is that kubelet won't evict other pods when there is a resource crunch just to accommodate this pod. Is this desired behaviour?

**Comment by dashpole** [comment]
@ravisantoshgudimetla yes, that is the desired behavior.  The kubelet doesn't treat static pods any differently than other pods.

**Comment by vainu-arto** [comment]
So @dashpole you are saying that the previous claim by @bsalamat that "This is WAI. If you enable Pod priority and preemption and some of your Pods have larger than zero priority, they could cause lower priority pods to get evicted." is incorrect, and there are multiple bugs here instead of just one?

Bug 1. is well known at this point (it's not possible to set a priority for a static pod).
Bug 2. is that critical annotation is ignored when priorities are in use
Bug 3. is that static pods are removed by something (either evicted or rescheduled) when they should not be

Am I now finally clued in?
